skll/learner.py

```diff
@@ -1465,56 +1533,55 @@ def train(self, examples, param_grid=None, grid_search_folds=3,
-                    # if all of the labels are strings, only class 1 objectives
-                    # are valid (with a classifier).
-                    raise ValueError("{} is not a valid grid objective "
-                                     "function for the {} learner with string "
-                                     "labels".format(grid_objective,
-                                                     self._model_type.__name__))
-
-                elif len(set(examples.labels)) == 2:
-                    # If there are two labels, class 3 objectives are valid for
-                    # classifiers regardless of the type of the label.
-                    if grid_objective not in _BINARY_CLASS_OBJ_FUNCS:
-                        raise ValueError("{} is not a valid grid objective "
-                                         "function for the {} learner with "
-                                         "binary labels"
-                                         .format(grid_objective,
-                                                 self._model_type.__name__))
-                elif grid_objective in _REGRESSION_ONLY_OBJ_FUNCS:
-                    # simple backoff check for mixed-type labels
-                    raise ValueError("{} is not a valid grid objective "
-                                     "function for the {} learner"
-                                     .format(grid_objective,
-                                             self._model_type.__name__))
+
+            # get the list of objectives that are acceptable in the current
+            # prediction scenario and raise an exception if the current
+            # objective is not in this allowed list
+            label_type = examples.labels.dtype.type
+            if estimator_type == 'classifier':
+                sorted_unique_labels = np.unique(examples.labels)
+                allowed_objectives = _get_acceptable_classification_metrics(sorted_unique_labels)
+            else:
+                allowed_objectives = _get_acceptable_regression_metrics()
+
+            if grid_objective not in allowed_objectives:
+                raise ValueError("'{}' is not a valid objective "
+                                 "function for {} with "
+                                 "labels of type {}.".format(grid_objective,
+                                                             self._model_type.__name__,
+                                                             label_type.__name__))
+
+            # If we're using a correlation metric for doing binary
+            # classification and probability is set to true, we assume
+            # that the user actually wants the `_with_probabilities`
+            # version of the metric
+            if (grid_objective in _CORRELATION_METRICS and
+                    estimator_type == 'classifier' and
+                    self.probability):
+                self.logger.info('You specified "{}" as the objective with '
```

Comment:
This is a nice message and I confirmed that I could get it when expected (after making the abovementioned change):
```
2019-10-14 14:06:53,666 - Iris_CV_example_iris_MultinomialNB - INFO - You specified "spearman" as the objective with "probability" set to "true". If this is a binary classification task with integer labels, the probabilies for the positive class will be used to compute the correlation.
```

However, this statement does get logged a lot each time `train` is called. If you're doing cross-validation, it happens as many times as there are folds. I wonder if something can be done about that.

Also, due to the way it's being called right before the training is done, it's possible that somebody could write a config file that has multiple objectives and it goes through almost all the experiments and then fails on the last (say, because you are using `spearman`). I don't see an easy way of sort moving this check "up". But it would be nice if possible to fail early.

I think these two things are related in some way: if you could fail early, you wouldn't need to log it for each fold. You would just log it at that earlier point once.

Reply:
It would be nice to have these but I don't see a way of doing this without adding a lot more complexity and I don't think the annoyance of failing late or seeing multiple log messages warrants that. 

From another point of view, I'd say that this is just the price you pay for using the more convenient option of the config file. For example, we also read in the training file for every single sub-experiment if you have multiple learners and objectives. If these things annoy someone, they can just use the API directly.

Reply:
It would be nice to have these but I don't see a way of doing this without adding a lot more complexity and I don't think the annoyance of failing late or seeing multiple log messages warrants that.

From another point of view, I'd say that this is just the price you pay for using the more convenient option of the config file. For example, we also read in the training file for every single sub-experiment if you have multiple learners and objectives. If these things annoy someone, they can just use the API directly.